Remove dropwizard-jackson dep from core

dimas-b opened the pull request:
This is a simplification / cleanup. The dependency does not appear to be required in `polaris-core`

### How Has This Been Tested?

Manual `bootstrap` execution with EclipseLink and PostgreSQL.

adutra commented on polaris-core/src/main/java/org/apache/polaris/core/persistence/MetaStoreManagerFactory.java:
```diff
@@ -33,7 +32,7 @@
  * configuration
  */
 @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
-public interface MetaStoreManagerFactory extends Discoverable {
```
Interesting, isn't this going to disturb Dropwizard's so-called ["polymorphic configuration"](https://www.dropwizard.io/en/release-3.0.x/manual/configuration.html#polymorphic-configuration)?

It seems we rely heavily on that feature though:

https://github.com/apache/polaris/blob/7b2cca4add23e7be369971774a55ce3dae7cca14/polaris-service/src/main/resources/META-INF/services/io.dropwizard.jackson.Discoverable

eric-maynard replied:
Indeed I think we cannot safely remove `Discoverable` here

dimas-b replied:
well, it is loaded only in one place on startup (as far as I can tell) and I did verify that call path manually. Jackson was able to deserialize config and the store impl. correctly.

dimas-b replied:
FWIW, listing the factory class in `META-INF/services/io.dropwizard.jackson.Discoverable` _is_ important, but implementing `Discoverable` is not.

dimas-b replied:
the other relevant service declaration is https://github.com/apache/polaris/blob/8cb6b44bf57dc597dab612d109d3eb534aef5715/polaris-service/src/main/resources/META-INF/services/org.apache.polaris.core.persistence.MetaStoreManagerFactory#L21

dimas-b replied:
@eric-maynard : what other test can validate correct operation without implementing `Discoverable`?

sfc-gh-emaynard replied:
My understanding is that removing `Discoverable` from this will break any implementations that want to supply their own configs (I think that's the polymorphic configuration mentioned above?). 

If you can provide such an implementation in a test and the configs still get picked up, I think we would be good to merge

eric-maynard replied:
My understanding is that removing Discoverable from this will break any implementations that want to supply their own configs (I think that's the polymorphic configuration mentioned above?).

If you can provide such an implementation in a test and the configs still get picked up, I think we would be good to merge

dimas-b replied:
@eric-maynard : I've added `PolarisApplicationConfigurationTest` in the latest commit. Please have a look.

dimas-b replied:
@eric-maynard : WDYT?

eric-maynard replied:
The changes look good to me, but the test doesn't include any configs specific to that implementation.

dimas-b replied:
The new test in this PR demonstrates config polymorphism between two different implementations (coming from different modules/jars). @eric-maynard : could you give a bit more details on what you'd like to be added to tests?

adutra replied:
I think he meant to enhance the test with some fields populated through configuration:

```java
  @Nested
  class EclipseLinkMetastore {
    private final DropwizardAppExtension<PolarisApplicationConfig> app =
        new DropwizardAppExtension<>(
            PolarisApplication.class,
            CONFIG_PATH,
            RANDOM_APP_PORT,
            RANDOM_ADMIN_PORT,
            ConfigOverride.config("metaStoreManager.type", "eclipse-link"),
            ConfigOverride.config("metaStoreManager.persistence-unit", "test"),
            ConfigOverride.config("metaStoreManager.conf-file", "/path/to/persistence.xml"));

    @Test
    void testMetastoreType() {
      assertThat(app.getConfiguration().getMetaStoreManagerFactory())
          .asInstanceOf(type(EclipseLinkPolarisMetaStoreManagerFactory.class))
          .extracting("persistenceUnitName", "confFile")
          .containsExactly("test", "/path/to/persistence.xml");
    }
  }
```

(you need to switch the eclipse-link dependency from `testRuntimeOnly` to `testImplementation`)

dimas-b replied:
ok, will do

dimas-b replied:
added

dimas-b commented:
I'm going to add one more commit and rebase presently... hopefully clarifying the use of `Discoverable`.

dimas-b commented:
I rebased and refactored service descriptions (resources) to be coherent both with java type hierarchies and comply with the Dropwizard way of discovering Jackson sub-types.

As a side benefit, the last commit fixes this message that happens when running without EclipseLink:
```
INFO  [2024-11-12 19:41:53,334 - 12224 ] [main] [] i.d.j.DiscoverableSubtypeResolver: Unable to load org.apache.polaris.extension.persistence.impl.eclipselink.EclipseLinkPolarisMetaStoreManagerFactory 
```

adutra reviewed (approved):
LGTM – I think you achieved a nice balance between removing DW from core but keeping `Discoverable` around in relevant modules. 👍

collado-mike commented on extension/persistence/eclipselink/src/main/java/org/apache/polaris/extension/persistence/impl/eclipselink/DiscoverableMetaStoreManagerFactory.java:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @see DiscoverableSubtypeResolver
+ */
+public interface DiscoverableMetaStoreManagerFactory extends Discoverable {}
```
This doesn't make a lot of sense to me. When the subtype resolver is asked for implementations, it'll be for a common interface - i.e., `MetaStoreManagerFactory` in this case. Why would the subtype resolver ever ask for the eclipse-link specific version of the `DiscoverableMetaStoreManagerFactory`?

dimas-b replied:
Type resolution during YAML parsing is handled by Jackson, and it properly recognizes all types in the java hierarchy (even those not implementing `Discoverable`).

However, Jackson (in this setup) relies on `DiscoverableSubtypeResolver` to discover implementation classes. That latter part has peculiar logic, where `DiscoverableSubtypeResolver` performs a two stage lookup: 1) classes listed in `META-INF/services/io.dropwizard.jackson.Discoverable`, 2) classes listed in the service manifests for the classes from step 1. I do not really know why it does not directly inject classes from step 1 into Jackson.

I agree, it looks odd. However, having one `Discoverable` interface is not possible as this PR specifically proposes to remove DW dependencies from `polaris-core`, where the base interface exists and there is not other common root between the "in memory" metastore implementation and the EclipseLink implementation. We could add a rather small common module for that, of course, but in my mind having two separate `Discoverable` interfaces was less intrusive. WDYT?

eric-maynard replied:
> Leaf classes cannot implement {@link Discoverable} directly, they have to implement another interface, which implements {@link Discoverable}.

Isn't this the exact setup we have right now, where the leaf classes implement `MetastoreManagerFactory`?

Should we just move `MetastoreManagerFactory` out of core?

dimas-b replied:
> Should we just move MetastoreManagerFactory out of core?

That's an option. This is what I meant by "add a rather small common module" above, because the EclipseLink and In-Memory implementations have to share it.

Would do other people think? @collado-mike @adutra ?

adutra replied:
So, if I get this correctly, we would all love to have `MetastoreManagerFactory` directly in polaris-service like many other similar beans, but we can't, because of EclipseLink.

If that is correct, I'd prefer to keep your current solution _for now_, then revisit this later, when/if EclipseLink is replaced with some more robust persistence solution.

dimas-b commented:
Alternative implementation: #470
